### PR TITLE
Fabo/1873 big mergeroo

### LINKED ISSUE #1873: Merge web into develop

### PR DESCRIPTION
**Pull request by faboweb**
Closes #1873

_Description:_

- [ ] TmSessionImport shows some form validation issues
- [x] disable e2e test check on GitHub

❤️ Thank you!

---

- [ ] Added entries in `CHANGELOG.md` with issue # and GitHub username
- [ ] Reviewed `Files changed` in the github PR explorer

### CODE REVIEW

File: app/src/config.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "Cosmos Voyager",
+  "default_network": "local-testnet",
+  "denoms": ["stake", "photino"],
```
Comment (faboweb) [inline app/src/config.json]: instead of relying on a genesis, fixing the denoms for now

File: app/src/network.js [hunk 2]
```diff
@@ -1,19 +1,17 @@
...
-// COSMOS_NETWORK env var
-let DEFAULT_NETWORK = join(__dirname, `../networks/` + config.default_network)
-let networkPath = process.env.COSMOS_NETWORK || DEFAULT_NETWORK
+let networkPath = `../networks/` + config.default_network
 
-let genesisText = readFileSync(join(networkPath, `genesis.json`), `utf8`)
-let genesis = JSON.parse(genesisText)
-let networkName = genesis.chain_id
+export default async function() {
+  let genesis = (await axios(networkPath + `/genesis.json`)).data
```
Comment (fedekunze) [inline app/src/network.js]:
```suggestion
  let genesis = (await axios(`${networkPath}/genesis.json`)).data
```

File: app/src/network.js [hunk 3]
```diff
@@ -1,19 +1,17 @@
...
-let { join } = require(`path`)
-let { readFileSync } = require(`fs-extra`)
-let config = require(`./config.js`)
+let config = require(`./config.json`)
 
-// this network gets used if none is specified via the
-// COSMOS_NETWORK env var
-let DEFAULT_NETWORK = join(__dirname, `../networks/` + config.default_network)
-let networkPath = process.env.COSMOS_NETWORK || DEFAULT_NETWORK
+let networkPath = `../networks/` + config.default_network
```
Comment (fedekunze) [inline app/src/network.js]:
```suggestion
const networkPath = `../networks/${config.default_network}`
```

File: app/src/renderer/components/common/ShortBech32.vue [hunk 4]
```diff
@@ -25,7 +24,9 @@ export default {
     }
   },
   data: () => ({
-    showSuccess: false
+    showSuccess: false,
+    /* istanbul ignore next */
```
Comment (jbibla) [inline app/src/renderer/components/common/ShortBech32.vue]: istanbul?

Comment (faboweb) [reply app/src/renderer/components/common/ShortBech32.vue]: istanbul is responsible for code coverage. this tells it to exclude the following statement from the coverage report

File: app/src/renderer/components/common/TmSessionSignUp.vue [hunk 5]
```diff
@@ -1,6 +1,6 @@
 <template>
   <div class="tm-session">
-    <tm-form-struct :submit="() => onSubmit()" class="tm-session-container">
+    <tm-form-struct :submit="onSubmit.bind(this)" class="tm-session-container">
```
Comment (sabau) [inline app/src/renderer/components/common/TmSessionSignUp.vue]: separate PR
We can try a new method
```
:submit="wrapper(onSubmit)"

wrapper(func) {func(this)}
```

File: app/src/renderer/connectors/lcdClientMock.js [hunk 6]
```diff
@@ -357,7 +357,7 @@ let state = {
       voting_start_time: `2018-11-23T13:29:24.66404Z`,
       voting_end_time: `2018-11-25T13:29:24.66404Z`,
       proposal_status: `Passed`,
-      tally_result: {
+      final_tally_result: {
```
Comment (sabau) [inline app/src/renderer/connectors/lcdClientMock.js]: let's double check why at some point it changes name (comes from old PR)

File: app/src/renderer/main.js [hunk 7]
```diff
@@ -4,68 +4,71 @@
...
+import axios from "axios"
 
 import App from "./App"
 import routes from "./routes"
-import Node from "./connectors/node"
-import Store from "./vuex/store"
-import AxiosProxy from "./scripts/axiosProxy"
+import _Node from "./connectors/node"
+import _Store from "./vuex/store"
+const _config = require(`../../src/config.json`)
```
Comment (sabau) [inline app/src/renderer/main.js]:
```suggestion
const _config = require(`../config.json`)
```

File: app/src/renderer/scripts/keystore.js [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+  return wallet
+}
+export function importKey(name, password, seed) {
+  const wallet = generateWalletFromSeed(seed)
+  addKey(wallet, name, password)
+
+  return wallet
+}
+
+// TODO needs proof reading
```
Comment (sabau) [inline app/src/renderer/scripts/keystore.js]: + comments for poor mortals

File: app/src/renderer/scripts/wallet.js [hunk 9]
```diff
@@ -0,0 +1,183 @@
+const bip39 = require(`bip39`)
+const bip32 = require(`bip32`)
+const bech32 = require(`bech32`)
+const secp256k1 = require(`secp256k1`)
+import sha256 from "crypto-js/sha256"
+import ripemd160 from "crypto-js/ripemd160"
+import CryptoJS from "crypto-js"
+
```
Comment (sabau) [inline app/src/renderer/scripts/wallet.js]: maybe we can add some comment to explain the pieces of this string

File: tasks/gaia.js [hunk 10]
```diff
@@ -100,21 +84,23 @@ async function makeValidator(
...
+  await sendTokens(mainSignInfo, `10stake`, address, chainId)
+  console.log(`Waiting for funds to delegate`)
   while (true) {
-    console.log(`Waiting for funds to delegate`)
     try {
       await sleep(1000)
       await getBalance(cliHome, address)
     } catch (error) {
-      console.error(error) // kept in here to see if something unexpected fails
+      // console.error(error) // kept in here to see if something unexpected fails
```
Comment (fedekunze) [inline tasks/gaia.js]: Delete ?

File: tasks/gaia.js [hunk 11]
```diff
@@ -191,7 +177,7 @@ function startLocalNode(
   nodeOneId = ``
 ) {
   return new Promise((resolve, reject) => {
-    let command = `${nodeBinary} start --home ${nodeHome}` // TODO add --minimum_fees 1STAKE here
+    let command = `${nodeBinary} start --home ${nodeHome}` // TODO add --minimum_fees 1stake here
```
Comment (fedekunze) [inline tasks/gaia.js]: Create tracking issue for this `TODO`

File: app/src/renderer/vuex/modules/blockchain.js [hunk 12]
```diff
@@ -43,18 +43,9 @@ export default ({ node }) => {
...
-            (error, data) => {
-              if (error) {
-                reject(new Error(`Couldn't query block. ${error.message}`))
-              } else {
-                resolve(data.block_metas && data.block_metas[0])
-              }
-            }
-          )
-        })
+        blockMetaInfo = await node.rpc
```
Comment (sabau) [inline app/src/renderer/vuex/modules/blockchain.js]: Opinion (personal): usually I like to use only one way of handling promises

File: app/src/renderer/vuex/modules/blockchain.js [hunk 13]
```diff
@@ -79,13 +70,7 @@ export default ({ node }) => {
...
       commit(`setSubscribedRPC`, node.rpc)
 
-      function handleError(error) {
-        dispatch(`nodeHasHalted`)
-        state.error = error
-      }
-
-      node.rpc.status((error, status) => {
-        if (error) return handleError(error)
+      node.rpc.status().then(status => {
```
Comment (sabau) [inline app/src/renderer/vuex/modules/blockchain.js]: is it worth to convert subscribeToBlocks to async for the same reason as upper?

File: app/src/renderer/vuex/modules/delegation.js [hunk 14]
```diff
@@ -164,13 +165,15 @@ export default ({ node }) => {
       })
 
       // optimistic update the atoms of the user before we get the new values from chain
-      commit(`setAtoms`, user.atoms - amount)
+      commit(`updateWalletBalance`, {
```
Comment (sabau) [inline app/src/renderer/vuex/modules/delegation.js]: 👍

File: app/src/renderer/vuex/modules/delegation.js [hunk 15]
```diff
@@ -164,13 +165,15 @@ export default ({ node }) => {
...
+        denom,
+        amount: Number(liquidAtoms) - Number(amount)
+      })
       // optimistically update the committed delegations
-      Vue.set(
-        state.committedDelegates,
-        validator_addr,
-        state.committedDelegates[validator_addr] + amount
-      )
+      commit(`setCommittedDelegation`, {
```
Comment (sabau) [inline app/src/renderer/vuex/modules/delegation.js]: 👍 ops was this

File: app/src/renderer/vuex/modules/send.js [hunk 16]
```diff
@@ -13,86 +20,84 @@ export default ({ node }) => {
...
+        from: rootState.wallet.address,
+        account_number: rootState.wallet.accountNumber, // TODO move into LCD?
+        chain_id: rootState.connection.lastHeader.chain_id,
+        gas: String(config.default_gas),
+        generate_only: true
+      }
+      args.base_req = requestMetaData
+
+      // extract type
+      let type = args.type || `send`
```
Comment (sabau) [inline app/src/renderer/vuex/modules/send.js]: I would deconstruct the two static fields + ...everything else in a third variable for lien 81

File: app/src/renderer/vuex/modules/wallet.js [hunk 17]
```diff
@@ -87,42 +85,25 @@ export default ({ node }) => {
...
-      }
+    async sendCoins(
+      { dispatch, commit, state },
+      { receiver, amount, denom, password }
+    ) {
+      await dispatch(`sendTx`, {
+        type: `send`,
+        password,
+        to: receiver,
+        amount: [{ denom, amount: amount.toString() }]
```
Comment (sabau) [inline app/src/renderer/vuex/modules/wallet.js]: Would be nice to have numbers, after readind the dispatch probably would just add unnecessary complexity

File: test/unit/specs/components/common/TmSessionImport.spec.js [hunk 18]
```diff
@@ -161,7 +161,7 @@ describe(`TmSessionImport`, () => {
   })
 
   it(`should not continue if creation failed`, async () => {
-    store.dispatch = jest.fn(() => Promise.resolve(null))
+    store.dispatch = jest.fn(() => Promise.reject(new Error(`Wrong password`)))
```
Comment (sabau) [inline test/unit/specs/components/common/TmSessionImport.spec.js]: ❤️

File: test/unit/specs/scripts/wallet.spec.js [hunk 19]
```diff
@@ -0,0 +1,237 @@
...
+  createSignature,
+  createSignMessage,
+  generateWalletFromSeed,
+  generateSeed,
+  generateWallet,
+  createSignedTx,
+  createBroadcastBody
+} from "renderer/scripts/wallet.js"
+
+describe(`Key Generation`, () => {
```
Comment (sabau) [inline test/unit/specs/scripts/wallet.spec.js]: Minor: if you want to move the definition of replicated strings on the top of the file would be easier to read / match equal addresses etc

File: app/src/renderer/components/common/TmSessionSignUp.vue (hunk 5 above)
Comment (faboweb) [reply app/src/renderer/components/common/TmSessionSignUp.vue]: https://github.com/cosmos/voyager/issues/1890

File: app/src/renderer/scripts/wallet.js (hunk 9 above)
Comment (faboweb) [reply app/src/renderer/scripts/wallet.js]: https://github.com/cosmos/voyager/issues/1891

File: app/src/renderer/vuex/modules/blockchain.js (hunk 12 above)
Comment (faboweb) [reply app/src/renderer/vuex/modules/blockchain.js]: done

File: tasks/gaia.js (hunk 11 above)
Comment (faboweb) [reply tasks/gaia.js]: https://github.com/cosmos/voyager/issues/1721

File: app/src/renderer/vuex/modules/send.js (hunk 16 above)
Comment (faboweb) [reply app/src/renderer/vuex/modules/send.js]: https://github.com/cosmos/voyager/issues/1828

File: test/unit/specs/scripts/wallet.spec.js (hunk 19 above)
Comment (faboweb) [reply test/unit/specs/scripts/wallet.spec.js]: we can do this together with #1891

**Pull request merged**